### PR TITLE
feat: remove transmission

### PR DESCRIPTION
**Pull request by Freyskeyd**
https://img.shields.io/badge/3%20PR-grpc%20over%20p2p-blue 

# Description
Previous PR: #333 

This PR aims to remove the old request/response protocol to replace it with a more dedicated protocol that support gRPC calls over p2p connection.

Instead of forging and sending message through p2p, we reuse the gRPC definition and mechanisms to apply it to p2p connection. One server implementation could both serve p2p or http2 requests.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+        write!(f, "{}", self.0)
+    }
+}
+
+type ChannelNegotiationFuture =
+    BoxFuture<'static, (Result<Channel, tonic::transport::Error>, RequestId, PeerId)>;
+
+/// gRPC behaviour for libp2p
+///
+/// That allows to open gRPC connections to peers and to accept incoming gRPC connections.
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:71]:
```suggestion
/// Allows opening gRPC connections to peers and to accept incoming gRPC connections.
```

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+            next_inbound_request_id: Arc::new(AtomicU64::new(0)),
+            pending_outbound_connections: HashMap::new(),
+            pending_events: VecDeque::new(),
+            pending_negotiated_channels: FuturesUnordered::new(),
+        }
+    }
+
+    /// Ask the behaviour to create a new outbound connection for the given peer.
+    ///
+    /// The return value is a [`OutboundConnection`] that can be used to check the status of the
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:111]:
```suggestion
    /// The return value is an [`OutboundConnection`] that can be used to check the status of the
```

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+                    request_id: *request_id,
+                    channel: channel.clone(),
+                }),
+                Some(Connection {
+                    id,
+                    address,
+                    request_id: Some(request_id),
+                    channel,
+                }) => {
+                    debug!("Peer already connected but no channel bound");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:140]: "bound"? Not sure I understand.

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+            address,
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
+
+        // TODO refactor this to better handle connection
+        // If there is no current established connection it means that it's the
+        // first connection with that peer
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:218]: Can you add a ticket for this TODO so we don't lose track?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        &mut self,
+        ConnectionClosed {
+            peer_id,
+            connection_id,
+            endpoint,
+            handler,
+            remaining_established,
+        }: ConnectionClosed<<Self as NetworkBehaviour>::ConnectionHandler>,
+    ) {
+        warn!("Connection {connection_id} closed with peer {peer_id}");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:248]: This is probably normal behaviour and a common occurrence. Better to use `debug`?

File: crates/topos-p2p/src/behaviour/grpc.rs [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+                }) = self.pending_outbound_connections.get(peer_id)
+                {
+                    info!("gRPC Outbound connection established with {peer_id}");
+                    self.pending_events.push_back(ToSwarm::NotifyHandler {
+                        peer_id: *peer_id,
+                        handler: libp2p::swarm::NotifyHandler::One(connection.id),
+                        event: *request_id,
+                    });
+
+                    debug!("Sending handler notif to connect");
```
Comment (dvdplm) [inline crates/topos-p2p/src/behaviour/grpc.rs:310]: Where is this send happening? Is this line in the right place?

**Review by dvdplm — COMMENTED** [review]
This is by no means a complete review... I'll be back.

**Review by dvdplm — APPROVED** [review]
Looks like an important cleanup.

**Comment by Freyskeyd** [comment]
> This is by no means a complete review... I'll be back.

Yes and the PR is still in draft

File: crates/topos-p2p/src/behaviour/grpc/error.rs [hunk 7]
```diff
@@ -6,8 +6,12 @@ use tokio::sync::oneshot;
 pub enum OutboundError {
     #[error("Unable to Dial")]
     DialFailure,
+    #[error("Peer doesn't supported protocol: {0}")]
```
Comment (hadjiszs) [inline crates/topos-p2p/src/behaviour/grpc/error.rs:9]:
```suggestion
    #[error("Peer doesn't support protocol: {0}")]
```

**Pull request merged**